log-levels.md

```diff
@@ -0,0 +1,203 @@
+
+Specifically, avoid situations like these:
+
+```swift
+debug: conection established [connection-id: 7]
+debug: connection closed unexpectedly [error: foobar] // BAD, the connection-id was dropped
+```
+
+On the second line, we don't know which connection had the error since the `connection-id` was dropped. Make sure to audit your logging code to ensure all relevant log statements carry neccessary correlation identifiers.
+
+#### Structured Logging (Semantic Logging)
+
+Libraries may want to embrace the structured logging style. 
+
+It is a fantastic pattern which makes consuming logs in automated systems, and through grepping and other means much easier and future proof.
+
+Consider the folowing not structured log statement:
+
+```swift
+// NOT structured logging style
+log.info("Accepted connection \(connection.id) from \(connection.peer), total: \(connections.count)")
+```
+
+It contains 4 peieces of information:
+
+- we accepted a connection,
+- this is it's string representation,
+- it is from this peer,
+- we currently have `connections.count` active connections.
+
+While this log statement contains all useful information that we meant to relay to end users, it is hard to visually and mechanically parse the detailed information it contains. For example, if we know connections start failing around the time when we reach a total of 100 concurrent connections, it is not trivial to find the specific log statement at which we hit this number. We would have to `grep 'total: 100'` for example, however perhaps there are many other `"total: "` strings present in all of our log systems.
+
+Instead, we can express the same information using the structured logging pattern, as follows:
+
+```swift
+log.info("Accepted connection", metadata: [
+  "connection.id": "\(connnection.id)",
+  "connection.peer": "\(connection.peer)",
+  "connections.total": "\(connections.count)"
+])
+
+// e.g. output:
+// <date> info [connection.id:?,connection.peer:?, connections.total:?] Accepted connection
+```
+
+which can be formatted, depending on the logging backend, slightly differently on various systems, but even in the simple string representation of such log, we'd be able to grep for `connections.total: 100` rather than having to guess the right string to grep for.
+
+Also, since the message now does not contain all that much "human readable wording", it is less prone to randomly change from "Accepted" to "We have accepted" or vice versa which could break alerting systems which are set up to parse and alert on specific log messages.
+
+Needless to say, structured logs are very useful in combination with [swift-distributed-tracing](https://github.com/apple/swift-distributed-tracing)'s `LoggingContext`, which automatically populates the metadata with any present trace information. Thanks for this all logs made in response to some specific request will automatically carry the same TraceID.
```

Comment:
```suggestion
Structured logs are very useful in combination with [swift-distributed-tracing](https://github.com/apple/swift-distributed-tracing)'s `LoggingContext`, which automatically populates the metadata with any present trace information. Thanks for this all logs made in response to some specific request will automatically carry the same TraceID.
```

I try to avoid things like "obviously", "needless to say" etc. in docs, I think it's a bit unfriendly to beginners who it might not be actually obvious to 🙂

Reply:
Right, thanks for spotting it!

Reply:
Random: recently picked up a book about writing style, it's been great and also of course calls out for trimming all such words... still working on it :-) Thanks again